architecture/2024-08-applicationdeployment-controller.md

```diff
@@ -0,0 +1,370 @@
+input commands and their corresponding output. Include a bicep/helm code 
+sample, if this proposal involves updates to that experience.
+-->
+
+**Sample Input:**
+<!--
+Provide a sample CLI command input and/or bicep/helm code.
+-->
+
+#### `env.bicep`
+```bicep
+extension radius
+
+param k8sNamespace string
+
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'env'
+  properties: {
+    compute: {
+      kind: 'kubernetes'
+      namespace: k8sNamespace
+    }
+  }
+}
+```
+
+#### `env.bicepparam`
+```bicep
+using './env.bicep'
+
+param k8sNamespace string = 'demo'
+```
+
+```
+> rad applicationdeployment generate -t env.bicep -p env.bicepparam -o env.yaml
+```
+
+**Sample Output:**
+<!--
+Provide a sample output for the inputs provided above.
+-->
+```
+> rad applicationdeployment generate -t env.bicep -p env.bicepparam -o env.yaml
+
+Generating ApplicationDeployment resource...
+ApplicationDeployment resource generated at env.yaml
+
+To apply the ApplicationDeployment resource, run:
+kubectl apply -f env.yaml
+```
+
+#### `env.yaml`
+```yaml
+kind: ApplicationDeployment
+apiVersion: radapp.io/v1alpha3
+metadata:
+  name: demo
+  namespace: demo
+spec:
+  template: |
+    extension radius
+
+    param k8sNamespace string
+
+    resource env 'Applications.Core/environments@2023-10-01-preview' = {
+      name: 'env'
+      properties: {
+        compute: {
+          kind: 'kubernetes'
+          namespace: k8sNamespace
+        }
+      }
+    }
+  parameters: |
+    using './env.bicep'
+
+    param k8sNamespace string = 'demo'
+```
+
+## Design
+
+### High Level Design
+<!--
+High level overview of the data flow and key components.
+
+Provide a high-level description, using diagrams as appropriate, and top-level
+explanations to convey the architectural/design overview. Don’t go into a lot
+of details yet but provide enough information about the relationship between
+these components and other components. Call out or highlight new components
+that are not part of this feature (dependencies). This diagram generally
+treats the components as black boxes. Provide a pointer to a more detailed
+design document, if one exists. 
+-->
+
+This design proposes the creation of two new components of the Radius control plane: the ApplicationDeployment reconciler (part of the `radius-controller`) and a new service that is responsible for compiling ARM JSON from Bicep manifests (`radius-bicep`).
+
+The `radius-controller` will be responsible for reconciling ApplicationDeployment resources. When an ApplicationDeployment resource is created or updated, the controller will read the Bicep template and parameters from the ApplicationDeployment resource and send them to the `radius-bicep` service. The `radius-bicep` service will compile the Bicep and bicepparam files into ARM JSON and return the ARM JSON to the `radius-controller`. The `radius-controller` will then send the ARM JSON payload to UCP to deploy the resources.
```

Comment:
Do we need a webhook? For example, a validating webhook that will validate the changes (create/update/delete) to the `ApplicationDeployment` resources before being picked up by the reconciler? We have a validating webhook for our Recipe resources btw, if you need an example.

Ref: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/.